Page.empty() is still broken

### Expected Behavior

an empty page should still be mappable. 

```
    void "Empty Page Can be mapped Mapped"() {
        when:
        Page<String> page = Page.empty()
        def copy = page.map { it -> it }
        copy.getTotalSize()

        then:
        notThrown(Throwable.class)
    }
```



### Actual Behaviour

It seems like some toes were dipped in the water around this issue
https://github.com/micronaut-projects/micronaut-data/pull/3142/files

but I'm guessing there's some additional context as to why it's set to null rather than 0. 
https://github.com/micronaut-projects/micronaut-data/blob/833e04b9e648864a313425419465771d478bfd9e/data-model/src/main/java/io/micronaut/data/model/Page.java#L100

```
    void "Broken Empty Page When Mapped"() {
        when:
        Page<String> page = Page.empty()
        def copy = page.map { it -> it }
        copy.getTotalSize()

        then:
        thrown(java.lang.IllegalStateException.class)
    }
```

```
Caused by: java.lang.IllegalStateException: Page does not contain total count. It is likely that the Pageable needs to be modified to request this information.
```

### Steps To Reproduce

test code provided above

### Environment Information

any

### Example Application

_No response_

### Version

Micronaut Data 4.9.4

--- a/data-model/src/main/java/io/micronaut/data/model/CursoredPage.java
+++ b/data-model/src/main/java/io/micronaut/data/model/CursoredPage.java
@@ -52,7 +52,7 @@ import java.util.stream.Collectors;
 @DefaultImplementation(DefaultCursoredPage.class)
 public interface CursoredPage<T> extends Page<T> {
 
-    CursoredPage<?> EMPTY = new DefaultCursoredPage<>(Collections.emptyList(), Pageable.unpaged(), Collections.emptyList(), null);
+    CursoredPage<?> EMPTY = new DefaultCursoredPage<>(Collections.emptyList(), Pageable.unpaged(), Collections.emptyList(), -1L);
 
     /**
      * @return Whether this {@link CursoredPage} contains the total count of the records
@@ -125,6 +125,9 @@ public interface CursoredPage<T> extends Page<T> {
      */
     @Override
     default @NonNull <T2> CursoredPage<T2> map(Function<T, T2> function) {
+        if (this == EMPTY) {
+            return (CursoredPage<T2>) EMPTY;
+        }
         List<T2> content = getContent().stream().map(function).collect(Collectors.toList());
         return new DefaultCursoredPage<>(content, getPageable(), getCursors(), hasTotalSize() ? getTotalSize() : null);
     }

--- a/data-model/src/main/java/io/micronaut/data/model/Page.java
+++ b/data-model/src/main/java/io/micronaut/data/model/Page.java
@@ -96,6 +96,9 @@ public interface Page<T> extends Slice<T> {
      */
     @Override
     default @NonNull <T2> Page<T2> map(Function<T, T2> function) {
+        if (this == EMPTY) {
+            return (Page<T2>) EMPTY;
+        }
         List<T2> content = getContent().stream().map(function).toList();
         return new DefaultPage<>(content, getPageable(), hasTotalSize() ? getTotalSize() : null);
     }

--- a/data-model/src/test/groovy/io/micronaut/data/model/PageSpec.groovy
+++ b/data-model/src/test/groovy/io/micronaut/data/model/PageSpec.groovy
@@ -199,6 +199,24 @@ class PageSpec extends Specification {
         !mappedCursoredPage.hasTotalSize()
     }
 
+    void "empty page can be mapped "() {
+        when:
+        Page<String> page = Page.empty()
+        def copy = page.map { it -> it }
+        copy.getTotalSize()
+
+        then:
+        notThrown(Throwable)
+
+        when:
+        CursoredPage<String> cursoredPage = CursoredPage.empty()
+        def newCopy = cursoredPage.map {it -> it }
+        newCopy.getTotalSize()
+
+        then:
+        notThrown(Throwable)
+    }
+
     @EqualsAndHashCode
     @ToString
     @Serdeable

--- a/src/main/docs/guide/azureCosmos/azureCosmosCriteriaSpecifications.adoc
+++ b/src/main/docs/guide/azureCosmos/azureCosmosCriteriaSpecifications.adoc
@@ -4,7 +4,7 @@ dependency:jakarta.persistence:jakarta.persistence-api[]
 
 To implement queries that cannot be defined at the compile-time Micronaut Data introduces api:data.repository.JpaSpecificationExecutor[] repository interface that can be used to extend your repository interface:
 
-snippet::example.PersonRepository[project-base="doc-examples/mongo-example",source="main" tags="repository",indent="0"]
+snippet::example.PersonRepository[project-base="doc-examples/azure-cosmos-example",source="main" tags="repository",indent="0"]
 
 Each method expects a "specification" which is a functional interface with a set of Criteria API objects intended to build a query programmatically.
 

